src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -603,18 +599,47 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 		}
 	}
 
-	/* reject queries which include CommonTableExpr */
+	/* reject queries which include CommonTableExpr which aren't routable */
 	if (queryTree->cteList != NIL)
 	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "common table expressions are not supported in distributed "
-							 "modifications",
-							 NULL, NULL);
+		ListCell *cteCell = NULL;
+		DeferredErrorMessage *cteError = NULL;
+
+		foreach(cteCell, queryTree->cteList)
+		{
+			CommonTableExpr *cte = (CommonTableExpr *) lfirst(cteCell);
+			Query *cteQuery = (Query *) cte->ctequery;
+			bool requiresMasterEvaluation;
+
+			if (cteQuery->commandType != CMD_SELECT)
+			{
+				return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
+									 "Router planner doesn't support non-select common table expressions.",
+									 NULL, NULL);
+			}
+
+			requiresMasterEvaluation = RequiresMasterEvaluation(cteQuery);
```

Comment:
Hmz, I was looking at how we handled function calls in updates with subqueries and it turns out there is a bug there:

```
# UPDATE test SET y = 3 WHERE x IN (SELECT x FROM test WHERE x = 3 AND y > random());
LOG:  issuing UPDATE public.test_102023 test SET y = 3 WHERE (x OPERATOR(pg_catalog.=) ANY (SELECT test_1.x FROM public.test_102023 test_1 WHERE ((test_1.x OPERATOR(pg_catalog.=) 3) AND ((test_1.y)::double precision OPERATOR(pg_catalog.>) '0.3421865622513'::double precision))))
```

i.e. we're evaluating the random() on the coordinator, while we should be re-evaluating it for every row on the worker (and actually, this should be recursively planned).

It seems we're primarily doing function evaluation in UPDATE/DELETE to support stable functions like now(), which are expected to return the same result throughout the lifetime of the query. The function evaluation logic does not distinguish between stable and volatile functions. The assumption is that UPDATE and DELETE queries do not contain volatile functions because those are rejected, but it seems we're neglecting to do so in subqueries.

Basically, we should do function evaluation if required, but we need to reject volatile functions in both CTEs and subqueries (and leave them to recursive planning).

Reply:
This seems like an issue to be looked into as a separate PR

Reply:
Agree that we can defer fixing it for subqueries, though we should figure out how to get it right in CTEs.